File: slinn_cxx/net/http/http_headers.cpp
#include "http_headers.hpp"
#include "../../utils.hpp"
#include <stdexcept>


#define WHITESPACE_CHARACTERS " \n\r\t\f\v"


std::string_view httpVersionToStdString(HttpVersion httpVersion) {
    switch (httpVersion) {
        case HttpVersion::H09: return "HTTP/0.9";
        case HttpVersion::H1:  return "HTTP/1.0";
        case HttpVersion::H11: return "HTTP/1.1";
        case HttpVersion::H2:  return "HTTP/2.0";
        case HttpVersion::H3:   return "HTTP/3.0";
    }
    return "unknown";
}

HttpVersion stdStringToHttpVersion(const std::string_view& str){
    if(str == "HTTP/0.9") return HttpVersion::H09;
    if(str == "HTTP/1.0") return HttpVersion::H1;
    if(str == "HTTP/1.1") return HttpVersion::H11;
    if(str == "HTTP/2.0") return HttpVersion::H2;
    if(str == "HTTP/3.0") return HttpVersion::H3;
    return HttpVersion::H11;
}

std::string_view httpProtocolToStdString(HttpProtocol httpProtocol) {
    switch (httpProtocol) {
        case HttpProtocol::HTTP: return "HTTP";
        case HttpProtocol::HTTPS:  return "HTTPS";
    }
    return "unknown";
}

HttpProtocol stdStringToHttpProtocol(const std::string_view& str){
    if(str == "HTTP") return HttpProtocol::HTTP;
    if(str == "HTTPS") return HttpProtocol::HTTPS;
    return HttpProtocol::HTTP;
}

std::string titleCase(std::string text) {
    bool new_word = true;
    for (char &c : text) {
        if (std::isspace(static_cast<unsigned char>(c))) {
            new_word = true;
        } else if (new_word) {
            c = std::toupper(static_cast<unsigned char>(c));
            new_word = false;
        } else {
            c = std::tolower(static_cast<unsigned char>(c));
        }
    }
    return text;
}

void toLowerCase(std::string& text){
    for (char &c : text) {
        c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    }
}

void toUpperCase(std::string& text){
    for (char &c : text) {
        c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));
    }
}

std::string join(const std::vector<std::string>& tokens, const std::string& delimeter){
    auto joined_view = tokens | std::views::join_with(delimeter);
    return std::string{joined_view.begin(), joined_view.end()};
}

std::vector<std::string> split(const std::string& str, const std::string& delimiter = " ") {
    std::vector<std::string> tokens;
    size_t start = 0;
    size_t end = str.find(delimiter);

    while (end != std::string::npos) {
        tokens.push_back(str.substr(start, end - start));
        start = end + delimiter.length();
        end = str.find(delimiter, start);
    }

    tokens.push_back(str.substr(start)); 
    return tokens;
}

template <typename T>
std::vector<T> slice(std::vector<T> vec, long long from, long long to){
    auto sliced_view = vec | std::views::drop(from) | std::views::take(to - 1);
    return std::vector<T>(sliced_view.begin(), sliced_view.end());
}

std::string_view lstrip(std::string_view s) {
    auto start = s.find_first_not_of(WHITESPACE_CHARACTERS);
    if (start == std::string_view::npos) {
        return "";
    }
    s.remove_prefix(start);
    return s;
}

std::string_view rstrip(std::string_view s) {
    auto end = s.find_last_not_of(WHITESPACE_CHARACTERS);
    if (end == std::string_view::npos) {
        return "";
    }
    s.remove_suffix(s.size() - 1 - end);
    return s;
}

std::string_view strip(std::string_view s) {
    return lstrip(rstrip(s));
}

template <typename K, typename V>
std::optional<V> get_optional(const std::unordered_map<K, V>& map, const K& key) {
    auto it = map.find(key);
    if (it != map.end()) {
        return it->second;
    }
    return std::nullopt;
}

std::string HttpHeaders::normalizeKey(std::string& key) const{
    if(key.starts_with(":")){
        toLowerCase(key);
    }else{
        key = titleCase(key);
    }
    key = strip(key);
    if(key == "Host") return ":authority";
    return key;
}

std::string HttpHeaders::normalizeValue(const py::object& value) const{
    return representate(value).attr("decode")().cast<std::string>();
}


HttpHeaders HttpHeaders::parse(const std::string& rawHttp){
    std::string_view http = rawHttp;

    size_t pos = http.find("\r\n");
    std::string_view _startLine = http.substr(0, pos);
    size_t methodPos =  _startLine.find(" ");
    size_t pathPos = _startLine.find(" ", methodPos + 1);

    HttpHeaders headers(
        stdStringToHttpVersion(_startLine.substr(pathPos, pos))
    );
    std::string _method = std::string(_startLine.substr(0, methodPos));
    toUpperCase(_method);
    headers._add(":method", _method);
    headers._add(":path", std::string(_startLine.substr(methodPos+1, pathPos-2)));

    size_t current_pos = pos + 2;
    while (current_pos < http.size()) {
        size_t next_line = http.find("\r\n", current_pos);
        if (next_line == std::string::npos) break;

        std::string_view line = http.substr(current_pos, next_line - current_pos);
        if (line.empty()) break;

        size_t colon_pos = line.find(':');
        if (colon_pos != std::string::npos) {
            std::string_view key = line.substr(0, colon_pos);
            std::string_view value = line.substr(colon_pos + 1);
            
            // Только здесь мы делаем одну аллокацию для ключа (если нужно)
            // или вообще работаем с view, если мы знаем, что ключ не изменится
            headers._add(std::string(strip(key)), std::string(strip(value)));
        }

        current_pos = next_line + 2;
    }

    return headers;
}

HttpHeaders::HttpHeaders(
    HttpVersion version, std::unordered_map<std::string, std::vector<py::object>> default_headers
){
    this->version = version;
    for (const auto& [key, values] : default_headers){
        for(const auto& value : values){
            add(key, value);
        }
    }
}

std::optional<std::string> HttpHeaders::method() const{
    return get(":method");
}
std::optional<std::string> HttpHeaders::scheme() const{
    return get(":scheme");
}
std::optional<std::string> HttpHeaders::authority() const{
    return get(":authority");
}
std::optional<std::string> HttpHeaders::path() const{
    return get(":path");
}
std::optional<HttpProtocol> HttpHeaders::protocol() const{
    if (__contains__(":protocol")) return stdStringToHttpProtocol(std::string(get(":protocol")));
    return std::nullopt;
}
std::optional<std::string> HttpHeaders::status() const{
    return get(":status");
}

std::string HttpHeaders::get(std::string key, const py::object& default_value) const{
    std::vector<std::string> _values = values(key, std::vector<py::object>{default_value});
    if(!_values.size()) return normalizeValue(default_value);
    return _values[0];
}
std::vector<std::string> HttpHeaders::values(std::string key, const std::vector<py::object>& default_values) const{
    std::vector<std::string> normalized_default_vaules;
    for(auto value : default_values){
        normalized_default_vaules.push_back(normalizeValue(value));
    }
    return get_optional(_data, normalizeKey(key)).value_or(normalized_default_vaules);
}

HttpHeaders HttpHeaders::add(std::string key, const py::object& value){
    _add(key, normalizeValue(value));
    return *this;
}
HttpHeaders HttpHeaders::add_many(std::unordered_map<std::string, std::vector<py::object>> headers){
    for(const auto& [_key, values] : headers){
        std::string key = _key;
        key = normalizeKey(key);
        std::vector<std::string> normalized_default_vaules;
        for(auto value : values){
            normalized_default_vaules.push_back(normalizeValue(value));
        }
        if(__contains__(key)){
            std::vector<std::string>& _values = _data[key];
            _values.reserve(_values.size() + normalized_default_vaules.size());
            _values.insert(_values.end(), normalized_default_vaules.begin(), normalized_default_vaules.end());
        }else{
            _data[key] = normalized_default_vaules;
        }
    }
    return *this;
}
void HttpHeaders::_add(std::string key, const std::string& value){
    key = normalizeKey(key);
    if(__contains__(key)){
        _data[key].push_back(value);
    }else{
        _data[key] = std::vector<std::string>{value};
    }
}
HttpHeaders HttpHeaders::set(std::string key, const py::object& value){
    _data[normalizeKey(key)] = std::vector<std::string>{normalizeValue(value)};
    return *this;
}
HttpHeaders HttpHeaders::del(std::string key){
    _data.erase(key);
    return *this;
}
std::string HttpHeaders::pop(std::string key, py::ssize_t index){
    std::string back = _data[key].back();
    _data[key].pop_back();
    return back;
}
std::vector<std::string> HttpHeaders::keys() const{
    std::vector<std::string> _keys;
    _keys.reserve(_data.size());

    for (const auto& [key, value] : _data) {
        _keys.push_back(key);
    }
    return _keys;
}
bool HttpHeaders::__contains__(std::string key) const{
    return _data.contains(key);
}
py::bytes HttpHeaders::make() const{
    if(!status()) throw std::runtime_error("pseudo header :status is not provided");
    std::string http = std::string(httpVersionToStdString(version)) + " " + std::string(*status()) + "\r\n";
    for (const auto& [key, values] : _data) {
        if(key.starts_with(":")) continue;
        for(const auto& value : values){
            http.append(std::string(key) + ": " + std::string(value) + "\r\n");
        }
    }
    http.append("\r\n");
    std::string_view view(http);
    return py::bytes(view.data(), view.size());
}


HttpHeaders HttpHeaders::extend(const HttpHeaders& headers){
    for (const auto& [key, values] : headers._data){
        auto [it, inserted] = _data.try_emplace(key, values);
        if(!inserted){
            std::vector<std::string>& _values = _data[key];
            _values.reserve(_values.size() + values.size());
            _values.insert(_values.end(), values.begin(), values.end());
        }
    }
    return *this;
}


std::string HttpHeaders::__request_str__() const{

}

File: slinn/net/http/http_headers.py
from typing import Any, SupportsIndex, Iterable
from slinn.net.http.exceptions import PseudoHeaderIsNotProvided
from slinn.utils import representate
import urllib.parse
import enum

class _HttpVersion(enum.Enum):
    H09 = 'HTTP/0.9'
    H1 = 'HTTP/1.0'
    H11 = 'HTTP/1.1'
    H2 = 'HTTP/2.0'
    H3 = 'HTTP/3.0'


class _HttpProtocol(enum.Enum):
    HTTP = 'HTTP'
    HTTPS = 'HTTPS'


class _HttpHeaders:
    __slots__ = ('_data', 'version')

    @staticmethod
    def _normalize_value(value: Any) -> str:
        return value if isinstance(value, str) else representate(value).decode()

    @staticmethod
    def _normalize_key(key: str) -> str:
        key = key.lower() if key.startswith(':') else key.title()
        if key == 'Host':
            return ':authority'
        return key

    @staticmethod
    def parse(raw_http: bytes) -> _HttpHeaders:
        http = raw_http.split(b'\r\n')
        _method, *_path, _version = http[0].split(b' ')

        headers = _HttpHeaders(
            version = _HttpVersion(_version.decode()),
            default = {
                ':method': _method.decode(),
                ':path': urllib.parse.unquote_plus(b' '.join(_path).decode())
            }
        )

        for header in http[1:-2]:
            key, value = header.split(b':', 1)
            headers.add(key.strip().decode(), value.strip().decode())

        return headers

    def __init__(self, *, version: _HttpVersion = _HttpVersion.H11, default: str | None = None):
        self._data: dict[str, list[str]] = {}

        self.version: _HttpVersion = version
        if default:
            for key, value in default.items():
                self.add(key, value)

    @property
    def method(self) -> str | None:
        return self.get(':method').upper()

    @property
    def scheme(self) -> str | None:
        return self.get(':scheme')

    @property
    def authority(self) -> str | None:
        return self.get(':authority')

    @property
    def path(self) -> str | None:
        return self.get(':path')

    @property
    def protocol(self) -> _HttpProtocol | None:
        _protocol = self.get(':protocol')
        return _HttpProtocol(_protocol) if _protocol in _HttpProtocol else None

    @property
    def status(self) -> str | None:
        return self.get(':status')

    def get(self, key: str, default: Any = None) -> str:
        values = self.values(key, [default])
        return values[0] if values else default

    def values(self, key: str, default: Iterable[Any] | None = None) -> Iterable[str]:
        return self._data.get(_HttpHeaders._normalize_key(key), [_HttpHeaders._normalize_value(v) for v in default or []])

    def add(self, key: str, value: Any) -> _HttpHeaders:
        key = _HttpHeaders._normalize_key(key)
        value = _HttpHeaders._normalize_value(value)
        if key in self._data and isinstance(self._data[key], list):
            self._data[key].append(value)
        else:
            self._data[key] = [value]
        return self

    def set(self, key: str, value: Any) -> _HttpHeaders:
        self._data[_HttpHeaders._normalize_key(key)] = [_HttpHeaders._normalize_value(value)]
        return self

    def delete(self, key: str) -> _HttpHeaders:
        del self._data[_HttpHeaders._normalize_key(key)]
        return self

    def pop(self, key: str, index: SupportsIndex = -1) -> str:
        return self._data[_HttpHeaders._normalize_key(key)].pop(index)

    def keys(self):
        return self._data.keys()

    def __contains__(self, key: str):
        return _HttpHeaders._normalize_key(key) in self.keys()

    def make(self):
        if self.status is None:
            raise PseudoHeaderIsNotProvided(':status')
        http: list[bytes] = []
        http.extend((self.version.value.encode(), b' ', self.status.encode() + b'\r\n'))
        for key in self.keys():
            if key.startswith(':'):
                continue
            for value in self.values(key):
                http.extend((key.encode(), b': ', value.encode(), b'\r\n'))
        http.append(b'\r\n')
        return b''.join(http)


from slinn_cxx import HttpHeaders, HttpVersion, HttpProtocol

#HttpHeaders = _HttpHeaders
#HttpVersion = _HttpVersion
#HttpProtocol = _HttpProtocol

if __name__ == '__main__':
    import time
    count = 500000

    _start = time.time()
    for i in range(500000):
        HttpHeaders.parse(b"GET / HTTP/1.1\r\nHost: miot.su\r\n\r\n")
    print(f'C++ parsing {count} times', time.time() - _start)

    _start = time.time()
    for i in range(500000):
        _HttpHeaders.parse(b"GET / HTTP/1.1\r\nHost: miot.su\r\n\r\n")
    print(f'Python parsing {count} times', time.time() - _start)

    headers = HttpHeaders()
    headers.set(':status', '200 OK')
    headers.set('Server', 'Slinn')
    _start = time.time()
    for i in range(500000):
        headers.make()
    print(f'C++ making {count} times', time.time() - _start)

    _headers = _HttpHeaders()
    _headers.set(':status', '200 OK')
    _headers.set('Server', 'Slinn')
    _start = time.time()
    for i in range(500000):
        _headers.make()
    print(f'Python making {count} times', time.time() - _start)

